hal/halx86/apic/apictimer.c

```diff
@@ -63,22 +69,75 @@ VOID
 NTAPI
 HalStartProfileInterrupt(IN KPROFILE_SOURCE ProfileSource)
 {
-    UNIMPLEMENTED;
-    return;
+	LVT_REGISTER LvtEntry;
+
+	/* Only handle ProfileTime */
+	if (ProfileSource == ProfileTime)
+	{
+		/* OK, we are profiling now */
+		HalIsProfiling = TRUE;
+
+		/* Set interrupt interval */
+		ApicWrite(APIC_TICR, PCR->HalReserved[0]);
+
+		/* Unmask it */
+		LvtEntry.Long = 0;
+		LvtEntry.TimerMode = 1;
+		LvtEntry.Vector = APIC_PROFILE_VECTOR;
+		LvtEntry.Mask = 0;
+		ApicWrite(APIC_TMRLVTR, LvtEntry.Long);
+	}
 }
 
 VOID
 NTAPI
 HalStopProfileInterrupt(IN KPROFILE_SOURCE ProfileSource)
 {
-    UNIMPLEMENTED;
-    return;
+	LVT_REGISTER LvtEntry;
+
+	/* Only handle ProfileTime */
+	if (ProfileSource == ProfileTime)
+	{
+		/* We are not profiling */
+		HalIsProfiling = FALSE;
+
+		/* Mask interrupt */
+		LvtEntry.Long = 0;
+		LvtEntry.TimerMode = 1;
+		LvtEntry.Vector = APIC_PROFILE_VECTOR;
+		LvtEntry.Mask = 1;
+		ApicWrite(APIC_TMRLVTR, LvtEntry.Long);
+	}
 }
 
 ULONG_PTR
 NTAPI
 HalSetProfileInterval(IN ULONG_PTR Interval)
 {
-    UNIMPLEMENTED;
-    return Interval;
+	ULONGLONG TimerInterval;
+	ULONGLONG FixedInterval;
+
+	FixedInterval = (ULONGLONG)Interval;
+
+	/* Check bounds */
+	if (FixedInterval < HalMinProfileInterval)
+		FixedInterval = HalMinProfileInterval;
+	else if (FixedInterval > HalMaxProfileInterval)
+		FixedInterval = HalMaxProfileInterval;
+
+	/* Remember interval */
+	HalCurProfileInterval = FixedInterval;
+
+	/* Set multiplier to 1 */
+	PCR->HalReserved[1] = 1;
```

Comment:
Is this constant 1 a hack? If so, mark it as such please.

Reply:
More of this below.